File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.dart
class Solution {
  bool searchMatrix(List<List<int>> matrix, int target) {
    //? the number of rows

    int m = matrix.length;
    //? the number of column
    int n = matrix[0].length;
    int left = 0;
    int right = m * n - 1;
    while (left <= right) {
      int mid = (left + right) ~/ 2;
      //? the formula to extract the row index
      int i = mid ~/ n;
      //? the formula to extract the column index
      int j = mid % n;
      //? get the middle number index
      int mid_num = matrix[i][j];

      if (mid_num > target) {
        right = mid - 1;
      } else if (mid_num < target) {
        left = mid + 1;
      }
      if (mid_num == target) {
        return true;
      }
    }
    return false;
  }
}

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: list[list[int]], target: int) -> bool:
        m = len(matrix)
        # number of rows 
        n = len(matrix[0])
        t = m * n
        left = 0
        right = t - 1
        while(left <= right):
            mid = (left + right) // 2
            # formula to calculate the row index
            mid_i = mid // n
            # formula to calculate the column index
            mid_j = mid % n
            # always create a variable for the composed components
            mid_num = matrix[mid_i][mid_j]
            
            if(mid_num == target):
                return True
            elif(mid_num > target):
                right = mid - 1
            else:
                left = mid + 1
        return False